src/nvme/nbft.c

```diff
@@ -0,0 +1,947 @@
+};
+
+#define DISCOVERYFLAG_VALID	0x01
+
+/*
+ *  End of NBFT ACPI table definitions
+ */
+#define MIN(a,b) (((a)<(b))?(a):(b))
+
+static __u8 csum(void *buffer, int length)
+{
+	int n;
+	__u8 sum = 0;
+
+	for (n = 0; n < length; n++) {
+		sum = (__u8)(sum + ((__u8 *)buffer)[n]);
+	}
+	return sum;
+}
+
+static void format_ip_addr(char *buf, size_t buflen, __u8 *addr) {
+	struct in6_addr *addr_ipv6;
+
+	addr_ipv6 = (struct in6_addr *)addr;
+	if (   addr_ipv6->s6_addr32[0] == 0
+	    && addr_ipv6->s6_addr32[1] == 0
+	    && (ntohl(addr_ipv6->s6_addr32[2]) == 0xffff) )
+		/* ipv4 */
+		inet_ntop(AF_INET, &(addr_ipv6->s6_addr32[3]), buf, buflen);
+	else
+		/* ipv6 */
+		inet_ntop(AF_INET6, addr_ipv6, buf, buflen);
+}
+
+static int in_heap(struct raw_nbft_header *header, heap_obj obj)
+{
+	if (obj.length == 0)
+		return 1;
+	if (obj.offset < header->heap_offset)
+		goto bad;
+	if (obj.offset > header->heap_offset + header->heap_length)
+		goto bad;
+	if (obj.offset + obj.length > header->heap_offset + header->heap_length)
+		goto bad;
+	return 1;
+bad:
+	return 0;
+}
+
+/*
+ *  Return transport_type string (NBFT Table 2)
+ */
+static char *trtype_to_string(__u8 transport_type)
+{
+	switch (transport_type) {
+		case 3:
+			return "tcp";
+			break;
+		default:
+			return "invalid";
+			break;
+	}
+}
+
+#define verify(condition, message)							\
+	if (!(condition)) {								\
+		nvme_msg(NULL, LOG_DEBUG, "file %s: " message "\n", nbft->filename);	\
+		return -EINVAL;								\
+	}
+
+static int __get_heap_obj(struct raw_nbft_header *header, const char *filename,
+			  const char *descriptorname, const char *fieldname,
+			  heap_obj obj, bool is_string,
+			  char **output)
+{
+	if (obj.length == 0) {
+		*output = NULL;
+		return -ENOENT;
+	}
+
+	if (!in_heap(header, obj)) {
+		nvme_msg(NULL, LOG_DEBUG, "file %s: field '%s' in descriptor '%s' has invalid offset or length\n",
+			 filename, fieldname, descriptorname);
+		return -EINVAL;
+	}
+
+	/* check that string is zero terminated correctly */
+	*output = (char *)header + obj.offset;
+
+	if (is_string) {
+		if (strnlen(*output, obj.length + 1) < obj.length)
+			nvme_msg(NULL, LOG_DEBUG, "file %s: string '%s' in descriptor '%s' is shorter (%ld) than specified length (%d)\n",
+				filename, fieldname, descriptorname, strnlen(*output, obj.length + 1), obj.length);
+		else if (strnlen(*output, obj.length + 1) > obj.length) {
+			nvme_msg(NULL, LOG_DEBUG, "file %s: string '%s' in descriptor '%s' is not zero terminated\n",
+				 filename, fieldname, descriptorname);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+#define get_heap_obj(descriptor, obj, is_string, output)		\
+	__get_heap_obj(header, nbft->filename,			\
+		       stringify(descriptor), stringify(obj),	\
+		       descriptor->obj, is_string,				\
+		       output)
+
+static struct nbft_discovery *discovery_from_index(struct nbft_info *nbft, int i)
+{
+	struct nbft_discovery *d;
+
+	list_for_each(&nbft->discovery_list, d, node)
```

Comment:
Please add '{}' when it's more than one line.

Reply:
Not applicable anymore.